pkg/entitlements/service.go

```diff
@@ -0,0 +1,145 @@
+package entitlements
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/canonical/identity-platform-admin-ui/internal/logging"
+	"github.com/canonical/identity-platform-admin-ui/internal/monitoring"
+	"github.com/canonical/identity-platform-admin-ui/internal/pool"
+	"github.com/canonical/identity-platform-admin-ui/pkg/authentication"
+	v1 "github.com/canonical/rebac-admin-ui-handlers/v1"
+	"github.com/canonical/rebac-admin-ui-handlers/v1/resources"
+	"go.opentelemetry.io/otel/trace"
+)
+
+var ObjectTypes = [...]string{"role", "group", "identity", "scheme", "client", "provider"}
+
+type listEntitlementResult struct {
+	entitlementSchemas []resources.EntitlementSchema
+	err                error
+}
+
+type V1Service struct {
+	ofga  OpenFGAClientInterface
+	wpool pool.WorkerPoolInterface
+
+	tracer  trace.Tracer
+	monitor monitoring.MonitorInterface
+	logger  logging.LoggerInterface
+}
+
+func (s *V1Service) ListEntitlements(ctx context.Context, params *resources.GetEntitlementsParams) ([]resources.EntitlementSchema, error) {
+	ctx, span := s.tracer.Start(ctx, "entitlements.V1Service.ListEntitlements")
+	defer span.End()
+
+	principal := authentication.PrincipalFromContext(ctx)
+	if principal == nil {
+		return nil, v1.NewAuthorizationError("unauthorized")
+	}
+
+	results := make(chan *pool.Result[any], len(ObjectTypes))
+	wg := sync.WaitGroup{}
+	wg.Add(len(ObjectTypes))
+
+	for _, objectType := range ObjectTypes {
+		_, err := s.wpool.Submit(
+			s.listEntitlementsByObjectTypeFunc(ctx, fmt.Sprintf("user:%s", principal.Identifier()), objectType, *params.Filter),
```

Comment:
in this case instead of returning a proper tuple we should parse the schema and have 

`client`, `can_view`, `user,group#member,role#assignee`
`client`, `can_create`, `user,group#member,role#assignee`
`client`, `can_edit`, `user,group#member,role#assignee`

so no real tuples but the definition of the types

Reply:
Completely misunderstood what this API tries to do 😅. I will change it.